app/src/lib/stores/git-store.ts

```diff
@@ -689,6 +705,87 @@ export class GitStore {
     this.emitUpdate()
   }
 
+  /** Load the upstream remote if it exists. */
+  public async loadUpstreamRemote(): Promise<void> {
+    const parent =
+      this.repository.gitHubRepository &&
+      this.repository.gitHubRepository.parent
+    if (!parent) {
+      return
+    }
+
+    const remotes = await getRemotes(this.repository)
+    const upstream = this.findUpstreamRemote(parent, remotes)
+    this._upstream = upstream
+    this.emitUpdate()
+  }
+
+  /**
+   * Add the upstream remote if the repository is a fork and an upstream remote
+   * doesn't already exist.
+   */
+  public async addUpstreamRemoteIfNeeded(): Promise<void> {
+    const parent =
+      this.repository.gitHubRepository &&
+      this.repository.gitHubRepository.parent
+    if (!parent) {
+      return
+    }
+
+    const remotes = await getRemotes(this.repository)
+    const upstream = this.findUpstreamRemote(parent, remotes)
+    if (upstream) {
+      return
+    }
+
+    const remoteWithUpstreamName = remotes.find(
+      r => r.name === UpstreamRemoteName
+    )
+    if (remoteWithUpstreamName) {
+      const error = new UpstreamAlreadyExistsError(
+        this.repository,
+        remoteWithUpstreamName
+      )
+      this.emitError(error)
+      return
+    }
+
+    const url = forceUnwrap(
+      'Parent repositories are fully loaded',
+      parent.cloneURL
+    )
+
+    await this.performFailableOperation(() =>
+      addRemote(this.repository, UpstreamRemoteName, url)
+    )
+    this._upstream = { name: UpstreamRemoteName, url }
+  }
+
+  private findUpstreamRemote(
+    parent: GitHubRepository,
+    remotes: ReadonlyArray<IRemote>
+  ): IRemote | null {
+    const upstream = remotes.find(r => r.name === UpstreamRemoteName)
```

Comment:
What if the user named their remote something other than upstream?

Reply:
The same as if they didn't have it: we'd add it.